File: blue-core/src/test/java/blue/utility/UDOUtilitiesTest.java
package blue.utility;

import blue.udo.OpcodeList;
import blue.udo.UDOStyle;
import blue.udo.UserDefinedOpcode;
import electric.xml.Element;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;
import org.junit.jupiter.api.Test;

class UDOUtilitiesTest {

    @Test
    void parsesModernUDOText() {
        String text = """
                opcode saturate(aSig, kDrive):a
                aOut = tanh(aSig * kDrive)
                xout aOut
                endop
                """;

        OpcodeList udos = UDOUtilities.parseUDOText(text);

        assertEquals(1, udos.size());

        UserDefinedOpcode udo = udos.get(0);
        assertEquals(UDOStyle.MODERN, udo.style);
        assertEquals("saturate", udo.opcodeName);
        assertEquals("aSig, kDrive", udo.inputArguments);
        assertEquals("ak", UDOUtilities.getInTypesFromInputArguments(udo.inputArguments));
        assertEquals("a", udo.outTypes);
    }

    @Test
    void parsesModernUDOTextWithAnnotatedArgsAndSpacedColon() {
        String text = """
                opcode myOpcode(kIn1:o, kIn2:j) : a
                xout 0
                endop
                """;

        OpcodeList udos = UDOUtilities.parseUDOText(text);

        assertEquals(1, udos.size());

        UserDefinedOpcode udo = udos.get(0);
        assertEquals(UDOStyle.MODERN, udo.style);
        assertEquals("myOpcode", udo.opcodeName);
        assertEquals("kIn1:o, kIn2:j", udo.inputArguments);
        assertEquals("oj", UDOUtilities.getInTypesFromInputArguments(udo.inputArguments));
        assertEquals("a", udo.outTypes);
    }

    @Test
    void parsesModernUDOTextWhenColonStartsNextLine() {
        String text = """
                opcode nextLineColon(aSig, kDrive)
                : (a, a)
                xout aSig, aSig
                endop
                """;

        OpcodeList udos = UDOUtilities.parseUDOText(text);

        assertEquals(1, udos.size());

        UserDefinedOpcode udo = udos.get(0);
        assertEquals(UDOStyle.MODERN, udo.style);
        assertEquals("nextLineColon", udo.opcodeName);
        assertEquals("aSig, kDrive", udo.inputArguments);
        assertEquals("a, a", udo.outTypes);
    }

    @Test
    void recoversWhenBrokenModernHeaderIsFollowedByAnotherOpcode() {
        String text = """
                opcode broken(aSig)
                opcode next, a, a
                aSig\txin
                xout aSig
                endop
                """;

        OpcodeList udos = UDOUtilities.parseUDOText(text);

        assertEquals(1, udos.size());

        UserDefinedOpcode udo = udos.get(0);
        assertEquals(UDOStyle.CLASSIC, udo.style);
        assertEquals("next", udo.opcodeName);
        assertEquals("a", udo.outTypes);
        assertEquals("a", udo.inTypes);
    }

    @Test
    void parsesClassicUDOText() {
        String text = """
                opcode saturate, a, ak
                aSig, kDrive\txin
                aOut = tanh(aSig * kDrive)
                xout aOut
                endop
                """;

        OpcodeList udos = UDOUtilities.parseUDOText(text);

        assertEquals(1, udos.size());

        UserDefinedOpcode udo = udos.get(0);
        assertEquals(UDOStyle.CLASSIC, udo.style);
        assertEquals("a", udo.outTypes);
        assertEquals("ak", udo.inTypes);
    }

    @Test
    void convertsClassicToModernUsingXinArguments() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.CLASSIC;
        udo.opcodeName = "saturate";
        udo.inTypes = "ak";
        udo.outTypes = "a";
        udo.codeBody = """
                aSig, kDrive\txin
                aOut = tanh(aSig * kDrive)
                xout aOut
                """;

        UDOUtilities.convertToModern(udo);

        assertEquals(UDOStyle.MODERN, udo.style);
        assertEquals("aSig, kDrive", udo.inputArguments);
        assertEquals("", udo.inTypes);
        assertFalse(udo.codeBody.contains("xin"));
        assertEquals("ak", UDOUtilities.getInTypesFromInputArguments(udo.inputArguments));
        assertEquals("a", udo.outTypes);
    }

    @Test
    void convertsClassicToModernPreservesLegacyInputKinds() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.CLASSIC;
        udo.opcodeName = "legacyKinds";
        udo.inTypes = "oj";
        udo.outTypes = "a";
        udo.codeBody = """
                kIn1, kIn2\txin
                xout 0
                """;

        UDOUtilities.convertToModern(udo);

        assertEquals(UDOStyle.MODERN, udo.style);
        assertEquals("kIn1:o, kIn2:j", udo.inputArguments);
        assertEquals("", udo.inTypes);
    }

    @Test
    void convertsModernToClassic() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.MODERN;
        udo.opcodeName = "saturate";
        udo.inputArguments = "aSig, kDrive";
        udo.outTypes = "a";
        udo.codeBody = """
                aOut = tanh(aSig * kDrive)
                xout aOut
                """;

        UDOUtilities.convertToClassic(udo);

        assertEquals(UDOStyle.CLASSIC, udo.style);
        assertEquals("ak", udo.inTypes);
        assertEquals("a", udo.outTypes);
        assertTrue(udo.codeBody.startsWith("aSig, kDrive\txin"));
        assertEquals("", udo.inputArguments);
    }

    @Test
    void convertsAnnotatedModernToClassicWithoutKeepingAnnotationsInXin() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.MODERN;
        udo.opcodeName = "legacyKinds";
        udo.inputArguments = "kIn1:o, kIn2:j";
        udo.outTypes = "a";
        udo.codeBody = "xout 0\n";

        UDOUtilities.convertToClassic(udo);

        assertEquals(UDOStyle.CLASSIC, udo.style);
        assertEquals("oj", udo.inTypes);
        assertTrue(udo.codeBody.startsWith("kIn1, kIn2\txin"));
        assertFalse(udo.codeBody.startsWith("kIn1:o"));
    }

    @Test
    void savesLoadsAndGeneratesModernUDO() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.MODERN;
        udo.opcodeName = "stereo_width";
        udo.inputArguments = "aSig, kWidth";
        udo.outTypes = "a, a";
        udo.codeBody = """
                aDel = delay(aSig, kWidth)
                xout aSig, aDel
                """;

        UserDefinedOpcode copy = UserDefinedOpcode.loadFromXML(udo.saveAsXML());

        assertEquals(UDOStyle.MODERN, copy.style);
        assertEquals("aSig, kWidth", copy.inputArguments);
        assertEquals("", copy.inTypes);
        assertEquals("a, a", copy.outTypes);
        assertTrue(copy.generateCode().startsWith("opcode stereo_width(aSig, kWidth):(a,a)\n"));
    }

    @Test
    void parsedModernUDORemainsEquivalentAfterXmlRoundTrip() {
        UserDefinedOpcode parsed = UDOUtilities.parseUDOText("""
                opcode foo(kIn1:o, kIn2:j):a
                xout 0
                endop
                """).get(0);

        UserDefinedOpcode loaded = UserDefinedOpcode.loadFromXML(parsed.saveAsXML());

        assertTrue(loaded.isEquivalent(parsed));
    }

    @Test
    void modernEquivalentComparisonIgnoresOutTypeSpacing() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.MODERN;
        udo.opcodeName = "fx";
        udo.inputArguments = "ain1, ain2";
        udo.outTypes = "a, a";
        udo.codeBody = "xout ain1, ain2";

        UserDefinedOpcode parsed = new UserDefinedOpcode(udo);
        parsed.outTypes = "a,a";

        assertTrue(udo.isEquivalent(parsed));

        OpcodeList list = new OpcodeList();
        list.addOpcode(udo);
        assertEquals("fx", list.getNameOfEquivalentCopy(parsed));
    }

    @Test
    void parsesModernUDOWithVoidOutput() {
        String text = """
                opcode logMsg(SMsg):void
                prints SMsg
                endop
                """;

        OpcodeList udos = UDOUtilities.parseUDOText(text);
        assertEquals(1, udos.size());

        UserDefinedOpcode udo = udos.get(0);
        assertEquals(UDOStyle.MODERN, udo.style);
        assertEquals("logMsg", udo.opcodeName);
        assertEquals("", udo.outTypes);

        String code = udo.generateCode();
        assertTrue(code.contains("opcode logMsg(SMsg):void"), code);
    }

    @Test
    void parsesModernUDOWithEmptyOutputList() {
        String text = """
                opcode logMsg(SMsg):()
                prints SMsg
                endop
                """;

        OpcodeList udos = UDOUtilities.parseUDOText(text);
        assertEquals(1, udos.size());

        UserDefinedOpcode udo = udos.get(0);
        assertEquals(UDOStyle.MODERN, udo.style);
        assertEquals("", udo.outTypes);
        assertTrue(udo.generateCode().contains("opcode logMsg(SMsg):void"));
    }

    @Test
    void parsesModernUDOWithMultipleOutputTypes() {
        String text = """
                opcode stereoSplit(aSig, kPan):(a, a)
                aL = aSig * (1 - kPan)
                aR = aSig * kPan
                xout aL, aR
                endop
                """;

        OpcodeList udos = UDOUtilities.parseUDOText(text);
        assertEquals(1, udos.size());

        UserDefinedOpcode udo = udos.get(0);
        assertEquals(UDOStyle.MODERN, udo.style);
        assertEquals("a, a", udo.outTypes);
        assertEquals("ak", UDOUtilities.getInTypesFromInputArguments(udo.inputArguments));

        String code = udo.generateCode();
        assertTrue(code.contains("opcode stereoSplit(aSig, kPan):(a,a)"), code);
    }

    @Test
    void convertsClassicToModernWithMultipleOutputs() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.CLASSIC;
        udo.opcodeName = "stereo";
        udo.inTypes = "a";
        udo.outTypes = "aa";
        udo.codeBody = """
                aSig\txin
                xout aSig, aSig
                """;

        UDOUtilities.convertToModern(udo);

        assertEquals(UDOStyle.MODERN, udo.style);
        assertEquals("a, a", udo.outTypes);
    }

    @Test
    void convertsClassicZeroOutputToModernVoidGeneration() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.CLASSIC;
        udo.opcodeName = "logMsg";
        udo.inTypes = "S";
        udo.outTypes = "0";
        udo.codeBody = """
                SMsg\txin
                prints SMsg
                """;

        UDOUtilities.convertToModern(udo);

        assertEquals("", udo.outTypes);
        assertTrue(udo.generateCode().contains("opcode logMsg(SMsg):void"));
    }

    @Test
    void convertsModernVoidToClassicZero() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.MODERN;
        udo.opcodeName = "logMsg";
        udo.inputArguments = "SMsg";
        udo.outTypes = "";
        udo.codeBody = "prints SMsg\n";

        UDOUtilities.convertToClassic(udo);

        assertEquals(UDOStyle.CLASSIC, udo.style);
        assertEquals("0", udo.outTypes);
        assertEquals("S", udo.inTypes);
    }

    @Test
    void modernGenerateCodeUsesFlushHeaderAndIndentedBody() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.MODERN;
        udo.opcodeName = "logMsg";
        udo.inputArguments = "SMsg";
        udo.outTypes = "";
        udo.codeBody = """
                prints SMsg
                xout 0
                """;

        String code = udo.generateCode();

        assertEquals("""
                opcode logMsg(SMsg):void
                    prints SMsg
                    xout 0
                endop""", code);
    }

    @Test
    void savesModernVoidOutTypesAsVoidTextInXml() {
        UserDefinedOpcode udo = new UserDefinedOpcode();
        udo.style = UDOStyle.MODERN;
        udo.opcodeName = "logMsg";
        udo.inputArguments = "SMsg";
        udo.outTypes = "";
        udo.codeBody = "prints SMsg\n";

        Element xml = udo.saveAsXML();

        assertEquals("void", xml.getElement("outTypes").getTextString());

        UserDefinedOpcode loaded = UserDefinedOpcode.loadFromXML(xml);
        assertEquals("", loaded.outTypes);
        assertTrue(loaded.generateCode().contains("opcode logMsg(SMsg):void"));
    }

    @Test
    void modernOutputSignatureHandlesVoid() {
        assertEquals("void", UDOUtilities.getModernOutputSignature("void"));
        assertEquals("void", UDOUtilities.getModernOutputSignature("0"));
        assertEquals("void", UDOUtilities.getModernOutputSignature(""));
        assertEquals("void", UDOUtilities.getModernOutputSignature("()"));
        assertEquals("a", UDOUtilities.getModernOutputSignature("a"));
        assertEquals("(a,k)", UDOUtilities.getModernOutputSignature("a, k"));
    }

    @Test
    void modernOutTypesDisplayUsesVoidForZeroOutput() {
        assertEquals("void", UDOUtilities.getModernOutTypesDisplay("void"));
        assertEquals("void", UDOUtilities.getModernOutTypesDisplay("0"));
        assertEquals("void", UDOUtilities.getModernOutTypesDisplay(""));
        assertEquals("void", UDOUtilities.getModernOutTypesDisplay("()"));
        assertEquals("a, a", UDOUtilities.getModernOutTypesDisplay("a, a"));
    }
}

File: packages/blue-data/tests/udo/udo-utilities.test.ts
import { describe, it, expect } from 'vitest';
import {
  parseUDOText,
  parseUDODeclaration,
  convertToModern,
  convertToClassic,
} from '../../src/opcodes/udo-utilities';
import {
  getModernOutputSignature,
  getModernOutTypesDisplay,
  getInTypesFromInputArguments,
  normalizeModernOutTypes,
  normalizeClassicOutTypes,
} from '../../src/opcodes/udo-type-utils';
import { OpcodeDefinition } from '../../src/opcodes/opcode-definition';
import { OpcodeList } from '../../src/opcodes/opcode-list';
import { UDOStyle } from '../../src/opcodes/udo-style';
import { Element } from '../../src/serialization/xml-reader';

describe('UDOUtilities', () => {
  // ─── Parsing Modern UDO Text ───

  it('parses modern UDO text', () => {
    const text = `opcode saturate(aSig, kDrive):a
    aOut = tanh(aSig * kDrive)
    xout aOut
endop`;
    const list = parseUDOText(text);
    expect(list.size()).toBe(1);
    const udo = list.getOpcode(0)!;
    expect(udo.getStyle()).toBe(UDOStyle.MODERN);
    expect(udo.getName()).toBe('saturate');
    expect(udo.getInputArguments()).toBe('aSig, kDrive');
    expect(udo.getOutTypes()).toBe('a');
    expect(getInTypesFromInputArguments(udo.getInputArguments())).toBe('ak');
  });

  it('parses modern UDO with annotated args and spaced colon', () => {
    const text = `opcode annotated(kIn1:o, kIn2:j) : a
    xout kIn1 + kIn2
endop`;
    const list = parseUDOText(text);
    expect(list.size()).toBe(1);
    const udo = list.getOpcode(0)!;
    expect(udo.getStyle()).toBe(UDOStyle.MODERN);
    expect(udo.getInputArguments()).toBe('kIn1:o, kIn2:j');
    expect(udo.getOutTypes()).toBe('a');
    expect(getInTypesFromInputArguments(udo.getInputArguments())).toBe('oj');
  });

  it('parses modern UDO when colon starts next line', () => {
    const text = `opcode nextLineColon(aSig, kDrive)
: (a, a)
    aL = aSig
    aR = aSig * kDrive
    xout aL, aR
endop`;
    const list = parseUDOText(text);
    expect(list.size()).toBe(1);
    const udo = list.getOpcode(0)!;
    expect(udo.getStyle()).toBe(UDOStyle.MODERN);
    expect(udo.getInputArguments()).toBe('aSig, kDrive');
    expect(udo.getOutTypes()).toBe('a, a');
  });

  it('recovers when broken modern header is followed by another opcode', () => {
    const text = `opcode broken(
this is not a valid header
opcode next, a, a
    xout 1
endop`;
    const list = parseUDOText(text);
    expect(list.size()).toBe(1);
    const udo = list.getOpcode(0)!;
    expect(udo.getStyle()).toBe(UDOStyle.CLASSIC);
    expect(udo.getName()).toBe('next');
    expect(udo.getOutTypes()).toBe('a');
    expect(udo.getInTypes()).toBe('a');
  });

  // ─── Parsing Classic UDO Text ───

  it('parses classic UDO text', () => {
    const text = `opcode saturate, a, ak
    aSig, kDrive    xin
    aOut = tanh(aSig * kDrive)
    xout aOut
    endop`;
    const list = parseUDOText(text);
    expect(list.size()).toBe(1);
    const udo = list.getOpcode(0)!;
    expect(udo.getStyle()).toBe(UDOStyle.CLASSIC);
    expect(udo.getName()).toBe('saturate');
    expect(udo.getOutTypes()).toBe('a');
    expect(udo.getInTypes()).toBe('ak');
  });

  // ─── Classic → Modern Conversion ───

  it('converts classic to modern using xin arguments', () => {
    const udo = new OpcodeDefinition();
    udo.setStyle(UDOStyle.CLASSIC);
    udo.setName('test');
    udo.setOutTypes('a');
    udo.setInTypes('ak');
    udo.setCode('aSig, kDrive\txin\naOut = tanh(aSig * kDrive)\nxout aOut');

    convertToModern(udo);

    expect(udo.getStyle()).toBe(UDOStyle.MODERN);
    expect(udo.getInputArguments()).toBe('aSig, kDrive');
    expect(udo.getInTypes()).toBe('');
    // xin line should be removed from code body
    expect(udo.getCode()).not.toContain('xin');
    expect(udo.getCode()).toContain('aOut = tanh(aSig * kDrive)');
  });

  it('converts classic to modern preserving legacy input kinds', () => {
    const udo = new OpcodeDefinition();
    udo.setStyle(UDOStyle.CLASSIC);
    udo.setName('test');
    udo.setOutTypes('a');
    udo.setInTypes('oj');
    udo.setCode('kIn1, kIn2\txin\nxout kIn1 + kIn2');

    convertToModern(udo);

    expect(udo.getStyle()).toBe(UDOStyle.MODERN);
    expect(udo.getInputArguments()).toBe('kIn1:o, kIn2:j');
  });

  it('converts classic to modern with multiple outputs', () => {
    const udo = new OpcodeDefinition();
    udo.setStyle(UDOStyle.CLASSIC);
    udo.setOutTypes('aa');
    udo.setInTypes('a');
    udo.setCode('xout 1, 2');

    convertToModern(udo);

    expect(udo.getOutTypes()).toBe('a, a');
  });

  it('converts classic zero output to modern void', () => {
    const udo = new OpcodeDefinition();
    udo.setStyle(UDOStyle.CLASSIC);
    udo.setOutTypes('0');
    udo.setInTypes('S');
    udo.setCode('prints Smsg');

    convertToModern(udo);

    expect(udo.getOutTypes()).toBe('');
    const code = udo.generateCode();
    expect(code).toContain(':void');
  });

  // ─── Modern → Classic Conversion ───

  it('converts modern to classic', () => {
    const udo = new OpcodeDefinition();
    udo.setStyle(UDOStyle.MODERN);
    udo.setName('test');
    udo.setInputArguments('aSig, kDrive');
    udo.setOutTypes('a');
    udo.setCode('aOut = tanh(aSig * kDrive)\nxout aOut');

    convertToClassic(udo);

    expect(udo.getStyle()).toBe(UDOStyle.CLASSIC);
    expect(udo.getInTypes()).toBe('ak');
    expect(udo.getInputArguments()).toBe('');
    expect(udo.getCode()).toContain('aSig, kDrive\txin');
    expect(udo.getOutTypes()).toBe('a');
  });

  it('converts annotated modern to classic without keeping annotations in xin', () => {
    const udo = new OpcodeDefinition();
    udo.setStyle(UDOStyle.MODERN);
    udo.setName('test');
    udo.setInputArguments('kIn1:o, kIn2:j');
    udo.setOutTypes('a');
    udo.setCode('xout kIn1 + kIn2');

    convertToClassic(udo);

    expect(udo.getStyle()).toBe(UDOStyle.CLASSIC);
    expect(udo.getInTypes()).toBe('oj');
    // xin line should have bare names (no :o or :j)
    expect(udo.getCode()).toContain('kIn1, kIn2\txin');
    expect(udo.getCode()).not.toContain('kIn1:o');
  });

  it('converts modern void to classic zero', () => {
    const udo = new OpcodeDefinition();
    udo.setStyle(UDOStyle.MODERN);
    udo.setInputArguments('SMsg');
    udo.setOutTypes('');
    udo.setCode('prints SMsg');

    convertToClassic(udo);

    expect(udo.getStyle()).toBe(UDOStyle.CLASSIC);
    expect(udo.getOutTypes()).toBe('0');
    expect(udo.getInTypes()).toBe('S');
  });

  // ─── XML Round-Trip ───

  it('saves loads and generates modern UDO', () => {
    const udo = new OpcodeDefinition();
    udo.setName('stereo_width');
    udo.setStyle(UDOStyle.MODERN);
    udo.setInputArguments('aSig, kWidth');
    udo.setOutTypes('a, a');
    udo.setCode('aL = aSig\naR = aSig * kWidth\nxout aL, aR');

    const xml = udo.saveAsXML();
    const loaded = OpcodeDefinition.loadFromXML(xml);

    expect(loaded.getStyle()).toBe(UDOStyle.MODERN);
    expect(loaded.getInputArguments()).toBe('aSig, kWidth');
    expect(loaded.getOutTypes()).toBe('a, a');

    const code = loaded.generateCode();
    expect(code).toContain('opcode stereo_width(aSig, kWidth):(a,a)');
  });

  it('parsed modern UDO remains equivalent after XML round-trip', () => {
    const text = `opcode fx(aSig, kDrive):a
    aOut = tanh(aSig * kDrive)
    xout aOut
endop`;
    const list = parseUDOText(text);
    const original = list.getOpcode(0)!;

    const xml = original.saveAsXML();
    const loaded = OpcodeDefinition.loadFromXML(xml);

    expect(original.isEquivalent(loaded)).toBe(true);
  });

  // ─── Equivalence with Spacing ───

  it('modern equivalent comparison ignores outType spacing', () => {
    const list = new OpcodeList();
    const a = new OpcodeDefinition();
    a.setName('fx');
    a.setStyle(UDOStyle.MODERN);
    a.setInputArguments('aSig, kDrive');
    a.setOutTypes('a, a');
    a.setCode('xout 1');
    list.addOpcode(a);

    const b = new OpcodeDefinition();
    b.setName('fx2');
    b.setStyle(UDOStyle.MODERN);
    b.setInputArguments('aSig, kDrive');
    b.setOutTypes('a,a');
    b.setCode('xout 1');

    expect(list.getNameOfEquivalentCopy(b)).toBe('fx');
  });

  // ─── Void and Empty Output ───

  it('parses modern UDO with void output', () => {
    const text = `opcode logMsg(SMsg):void
    prints SMsg
endop`;
    const list = parseUDOText(text);
    const udo = list.getOpcode(0)!;
    expect(udo.getOutTypes()).toBe('');
    const code = udo.generateCode();
    expect(code).toContain('opcode logMsg(SMsg):void');
  });

  it('parses modern UDO with empty output list', () => {
    const text = `opcode logMsg(SMsg):()
    prints SMsg
endop`;
    const list = parseUDOText(text);
    const udo = list.getOpcode(0)!;
    expect(udo.getOutTypes()).toBe('');
    const code = udo.generateCode();
    expect(code).toContain('opcode logMsg(SMsg):void');
  });

  it('parses modern UDO with multiple output types', () => {
    const text = `opcode stereoSplit(aSig, kPan):(a, a)
    aL = aSig * (1 - kPan)
    aR = aSig * kPan
    xout aL, aR
endop`;
    const list = parseUDOText(text);
    const udo = list.getOpcode(0)!;
    expect(udo.getOutTypes()).toBe('a, a');
    const code = udo.generateCode();
    expect(code).toContain('opcode stereoSplit(aSig, kPan):(a,a)');
  });

  // ─── Output Signature ───

  it('modern output signature handles all cases', () => {
    expect(getModernOutputSignature('void')).toBe('void');
    expect(getModernOutputSignature('0')).toBe('void');
    expect(getModernOutputSignature('')).toBe('void');
    expect(getModernOutputSignature('()')).toBe('void');
    expect(getModernOutputSignature('a')).toBe('a');
    expect(getModernOutputSignature('a, k')).toBe('(a,k)');
  });

  // ─── Display String ───

  it('modern outTypes display uses void for zero output', () => {
    expect(getModernOutTypesDisplay('void')).toBe('void');
    expect(getModernOutTypesDisplay('0')).toBe('void');
    expect(getModernOutTypesDisplay('')).toBe('void');
    expect(getModernOutTypesDisplay('()')).toBe('void');
    expect(getModernOutTypesDisplay('a, a')).toBe('a, a');
  });

  // ─── Code Generation Formatting ───

  it('modern generate code uses flush header and indented body', () => {
    const udo = new OpcodeDefinition();
    udo.setName('logMsg');
    udo.setStyle(UDOStyle.MODERN);
    udo.setInputArguments('SMsg');
    udo.setOutTypes('');
    udo.setCode('prints SMsg');

    const code = udo.generateCode();
    expect(code).toContain('opcode logMsg(SMsg):void\n');
    expect(code).toContain('    prints SMsg');
    expect(code).toContain('\nendop');
    // No tab indentation
    expect(code).not.toContain('\t');
  });

  // ─── XML Void Serialization ───

  it('saves modern void outTypes as void text in XML', () => {
    const udo = new OpcodeDefinition();
    udo.setName('logMsg');
    udo.setStyle(UDOStyle.MODERN);
    udo.setInputArguments('SMsg');
    udo.setOutTypes('');
    udo.setCode('prints SMsg');

    const xml = udo.saveAsXML();
    // Check that XML contains <outTypes>void</outTypes>
    const outTypesElem = xml.getElement('outTypes');
    expect(outTypesElem?.getTextString()).toBe('void');

    // Reload and verify
    const loaded = OpcodeDefinition.loadFromXML(xml);
    expect(loaded.getOutTypes()).toBe('');
    expect(loaded.generateCode()).toContain(':void');
  });

  // ─── parseUDODeclaration ───

  it('parseUDODeclaration returns null for non-opcode lines', () => {
    expect(parseUDODeclaration('instr 1')).toBeNull();
    expect(parseUDODeclaration('xout 1')).toBeNull();
  });

  it('parseUDODeclaration parses classic header', () => {
    const udo = parseUDODeclaration('opcode test, a, ak');
    expect(udo).not.toBeNull();
    expect(udo!.getStyle()).toBe(UDOStyle.CLASSIC);
    expect(udo!.getName()).toBe('test');
    expect(udo!.getOutTypes()).toBe('a');
    expect(udo!.getInTypes()).toBe('ak');
  });

  it('parseUDODeclaration parses modern header', () => {
    const udo = parseUDODeclaration('opcode saturate(aSig, kDrive):a');
    expect(udo).not.toBeNull();
    expect(udo!.getStyle()).toBe(UDOStyle.MODERN);
    expect(udo!.getName()).toBe('saturate');
    expect(udo!.getInputArguments()).toBe('aSig, kDrive');
    expect(udo!.getOutTypes()).toBe('a');
  });
});
